modules/friendlytag.js

```diff
@@ -1711,18 +1761,32 @@ Twinkle.tag.callbacks = {
 			summaryText += ' {{[[:' + (tagName.indexOf(':') !== -1 ? tagName : 'Template:' + tagName + '|' + tagName) + ']]}}';
 		};
 
-		if (!tags.length) {
-			Morebits.status.warn('Info', 'No tags remaining to apply');
-		}
-
 		tags.sort();
 		$.each(tags, addTag);
+		if (rcatsToAdd.length && rcatsToRemove.length) {
+			summaryText += ' & ';
+		}
+		rcatsToRemove.forEach(function removeTag(tag) {
+			var tag_re = new RegExp('\\{\\{' + Morebits.pageNameRegex(tag) + '\\s*(\\|[^}]+)?\\}\\}\\n?');
+
+			if (tag_re.test(pageText)) {
+				pageText = pageText.replace(tag_re, '');
+			}
+			summaryText += 'Removed {{[[:' + (tag.indexOf(':') !== -1 ? tag : 'Template:' + tag + '|' + tag) + ']]}}';
+		});
+		if (!tags.length && rcatsToRemove.length === 0) {
+			Morebits.status.warn('Info', 'No tags remaining to apply');
+		}
 
 		// Check for all Rcat shell redirects (from #433)
 		if (pageText.match(/{{(?:redr|this is a redirect|r(?:edirect)?(?:.?cat.*)?[ _]?sh)/i)) {
 			// Regex inspired by [[User:Kephir/gadgets/sagittarius.js]] ([[Special:PermaLink/831402893]])
 			var oldTags = pageText.match(/(\s*{{[A-Za-z\s]+\|(?:\s*1=)?)((?:[^|{}]*|{{[^}]*}})+)(}})\s*/i);
-			pageText = pageText.replace(oldTags[0], oldTags[1] + tagText + oldTags[2] + oldTags[3]);
+			if (rcatsToAdd.length) {
+				pageText = pageText.replace(oldTags[0], oldTags[1] + tagText + oldTags[2] + oldTags[3]);
+			} else {
+				pageText = pageText.replace(oldTags[0], '');
```

Comment:
This just blindly removes all tags if we're not adding any, right?  That's not what we want!  I think the goal here was to remove the rcat shell if there were no tags left (which, heh, I guess this technically does...) but that should be handled separately.